arcgis-ios-sdk-samples/Scenes/Animate image overlay/AnimateImagesWithImageOverlayViewController.swift

```diff
@@ -0,0 +1,235 @@
+                toolbar.items![index] = pauseButtonItem
+            } else {
+                let index = toolbar.items!.firstIndex(of: pauseButtonItem)!
+                toolbar.items![index] = playButtonItem
+            }
+        }
+    }
+    
+    // MARK: Initialize scene and animation
+    
+    /// Create a scene.
+    ///
+    /// - Returns: A new `AGSScene` object.
+    func makeScene() -> AGSScene {
+        // Create a tiled layer from World Dark Gray Base REST service.
+        let basemapTileURL = URL(string: "https://services.arcgisonline.com/arcgis/rest/services/Canvas/World_Dark_Gray_Base/MapServer")!
+        let worldDarkGrayBasemap = AGSArcGISTiledLayer(url: basemapTileURL)
+        // Create an elevation source from Terrain3D REST service.
+        let elevationServiceURL = URL(string: "https://elevation3d.arcgis.com/arcgis/rest/services/WorldElevation3D/Terrain3D/ImageServer")!
+        let elevationSource = AGSArcGISTiledElevationSource(url: elevationServiceURL)
+        // Create a surface and add it to the scene.
+        let surface = AGSSurface()
+        surface.elevationSources = [elevationSource]
+        let scene = AGSScene(basemap: AGSBasemap(baseLayer: worldDarkGrayBasemap))
+        scene.baseSurface = surface
+        return scene
+    }
+    
+    func startAnimation(fps: Int) {
+        // Invalidate display link to stop previous ongoing animation.
+        displaylink?.invalidate()
+        // Create a new display link.
+        displaylink = CADisplayLink(target: self, selector: #selector(setImageFrame))
+        // Set frame rate to 15, 30 or 60 frames per second.
+        displaylink!.preferredFramesPerSecond = fps
+        // Add to main thread common mode run loop, so it is not effected by UI events.
+        displaylink!.add(to: .main, forMode: .common)
+    }
+    
+    /// Make a circular iterator to provide image frames.
+    ///
+    /// - Returns: A circular iterator that returns the next `UIImage` everytime get called.
+    private func makeImagesIterator() -> CircularIterator<UIImage> {
+        // Get the URLs to those images added to the project's folder reference.
+        let pacificSouthWestURLs = Bundle.main.urls(forResourcesWithExtension: "png", subdirectory: "PacificSouthWest") ?? []
+        // Sort the image URLs by their relative pathnames.
+        let imageURLs = pacificSouthWestURLs.sorted { $0.lastPathComponent < $1.lastPathComponent }
+        return CircularIterator(
+            elements: imageURLs.map { UIImage(contentsOfFile: $0.path)! }
+        )
+    }
+    
+    /// Set current image to the image overlay.
+    @objc
+    func setImageFrame() {
+        let frame = AGSImageFrame(image: imagesIterator.next()!, extent: pacificSouthwestEnvelope)
+        imageOverlay.imageFrame = frame
+    }
+    
+    func getOpacityPercentageString(percentage: Float) -> String {
+        return "Opacity\n" + percentageFormatter.string(from: NSNumber(value: percentage))!
```

Comment:
I don't know if one is more efficient than the other, but you can also cast to `NSNumber` (`percentage as NSNumber`) instead of going through the initializer. I suspect casting might be more efficient, but I don't really know for sure.

Reply:
This is interesting. 

My knowledge for type casting still lives with C/C++. I couldn't find the details about how Swift types cast to ObjC types (most posts suggest it is Apple's compiling "magic") after digging a bit. Nevertheless it makes sense that compile time type casting `as` should be faster than runtime `init`. So I'll adopt this change.